### PR TITLE
Only walk into nodes with directives when processing tree for nullable directives.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryNullableDirective/CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -2,137 +2,174 @@
...
+using System;
+using System.Collections.Generic;
 using Microsoft.CodeAnalysis.CodeStyle;
 using Microsoft.CodeAnalysis.CSharp.LanguageService;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 
-namespace Microsoft.CodeAnalysis.CSharp.Analyzers.RemoveUnnecessaryNullableDirective
+namespace Microsoft.CodeAnalysis.CSharp.Analyzers.RemoveUnnecessaryNullableDirective;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryNullableDirective/CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer.cs:13]: view with whitespace off.

File: src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryNullableDirective/CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -2,137 +2,174 @@
...
 
-                var compilationOptions = context.Compilation.Options;
-                var defaultNullableContext = ((CSharpCompilationOptions)compilationOptions).NullableContextOptions;
-                context.RegisterSyntaxTreeAction(context =>
-                {
-                    if (ShouldSkipAnalysis(context, compilationOptions, notification: null))
-                        return;
+            var compilationOptions = context.Compilation.Options;
+            var defaultNullableContext = ((CSharpCompilationOptions)compilationOptions).NullableContextOptions;
+            context.RegisterSyntaxTreeAction(context => ProcessSyntaxTree(compilationOptions, defaultNullableContext, context));
```
Comment (ToddGrun) [inline src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryNullableDirective/CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer.cs:44]: >ProcessSyntaxTree [](http://example.com/codeflow?start=56&length=17)

Can't ProcessSyntaxTree pull the compilationOptions and defaultNullableContext out of the context itself, allowing this closure to be static?

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryNullableDirective/CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer.cs:44]: yes

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryNullableDirective/CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryNullableDirective/CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer.cs]: actuall no.  i need the COmpilation, which isn't exposed from a syntax tree action.

Comment (ToddGrun) [reply src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryNullableDirective/CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer.cs]: bit of a shame these Register* methods don't take in a context to allow these actions to be static.

**Comment by sharwell** [comment]
@CyrusNajmabadi This feels like a huge step backwards for the readability of the analyzer. Why not fix `GetNextDirective` to perform this optimization internally?

**Pull request merged**